### ISSUE TITLE
Hang (probably type inference) in Julia 1.2.0rc2

### ISSUE THREAD

**Issue by wbhart** [JuliaLang/julia#32595]
We have been having a hang when tests run in AbstractAlgebra.jl that seems to be a new regression for Julia-1.2.0.

I have been able to get it down to a more or less minimal example (< 300 sloc). Finding lines of code that I can remove now is becoming difficult, so I think this is as far as I can get it without a lot of extra work.

To reproduce it, simply check out the inference_bug branch of this repository (into a directory AbstractAlgebra):

https://github.com/wbhart/InferenceBug1

It's a very heavily cut down version of our AbstractAlgebra.jl package.

```
using Pkg
Pkg.test("AbstractAlgebra")
```

will hang indefinitely before the test runs.

Please note that this is a pure Julia package and I have removed all functions that didn't seem to contribute to the hang.

**Comment by JeffBezanson** [comment]
I get
```
ERROR: LoadError: UndefVarError: elem_type not defined
Stacktrace:
 [1] #FreeModule#3(::Bool, ::Type{AbstractAlgebra.Generic.FreeModule}, ::AbstractAlgebra.Integers{BigInt}, ::Int64) at /home/jeff/src/julia/AbstractAlgebra/src/generic/FreeModule.jl:8
 [2] Type at ./none:0 [inlined]
 [3] #FreeModule#1 at /home/jeff/src/julia/AbstractAlgebra/src/AbstractAlgebra.jl:44 [inlined]
 [4] FreeModule(::AbstractAlgebra.Integers{BigInt}, ::Int64) at /home/jeff/src/julia/AbstractAlgebra/src/AbstractAlgebra.jl:44
 [5] test_module() at /home/jeff/src/julia/AbstractAlgebra/test/AbstractAlgebra-test.jl:14
 [6] top-level scope at /home/jeff/src/julia/AbstractAlgebra/test/runtests.jl:7
 [7] include at ./boot.jl:328 [inlined]
```

**Comment by JeffBezanson** [comment]
Sorry, that was on master. Hangs in 1.2.

**Comment by JeffBezanson** [comment]
Ah, looks like this might be similar to #32488. Should be fixed by #32509; backported (#32592).

**Comment by wbhart** [comment]
Thanks @JeffBezanson Master has been causing us problems for quite a few months now. Hopefully this will fix a range of similar issues we've been facing. We do know there is at least one more of these that also affects Julia-1.1 as well though. We'll try and get you a minimal case ASAP.
